### ISSUE TITLE
No config support for user-added CA certificates?

### ISSUE THREAD

**Issue by jgilfelt** [commonsguy/cwac-netsecurity#5]
The following `networkSecurityConfig` declaration

```xml
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <debug-overrides>
        <trust-anchors>
            <!-- Trust user added CAs while debuggable only -->
            <certificates src="user" />
        </trust-anchors>
    </debug-overrides>
</network-security-config>
```

which is valid in the API 24+ implementation, produces the following exception on earlier Android versions:

`com.commonsware.cwac.netsecurity.config.XmlConfigSource$ParserException: Unknown certificates src. Should be one of system|user|@resourceVal at: Binary XML file line #6`

I can see that the configuration for user-added certificate sources has [not been included in the backport](https://github.com/commonsguy/cwac-netsecurity/blob/master/netsecurity/src/main/java/com/commonsware/cwac/netsecurity/config/XmlConfigSource.java#L209-L210). 

Is there any particular reason for this? A developer may not want to embed user generated certificates for things like Charles Proxy in their app. In any case, the exception message above should probably be amended to state that a `user` value is not valid.

**Comment by commonsguy** [comment]
> Is there any particular reason for this?

[`UserCertificateSource`](https://android.googlesource.com/platform/frameworks/base/+/android-7.1.1_r4/core/java/android/security/net/config/UserCertificateSource.java) relies on internal methods, unavailable in the Android SDK.

Moreover, prior to Android 7.0, user certificates will be used any time that system certificates are used. I can't change that platform behavior.

> A developer may not want to embed user generated certificates for things like Charles Proxy in their app.

User certificates should work as they used to prior to Android 7.0, so long as you are allowing system certificates.

> In any case, the exception message above should probably be amended to state that a user value is not valid.

I need to work through the proper graceful degradation policy here. Off the cuff and IMHO, the ideal situation is:

- If you are using `withManifestConfig()` &mdash; and so the backport is only used on Android 6.0 and older &mdash; `user` should not crash the app, but instead log a warning and behave the same as `system`

- If you are using anything else &mdash; and so the backport is used 100% of the time &mdash; `user` should crash the app with a better error message, to indicate that your configuration will not do what you think it does

Thoughts?

Thanks for pointing this out!

**Comment by jgilfelt** [comment]
Thanks for the clarification! I think the policies you suggest make perfect sense, and I can work around the crash for now by using a separate `xml-v24` resource. 

I still cannot for the life of me get my user certificate for Charles Proxy trusted alongside a custom `<domain-config>` on pre-Nougat devices, even when explicitly allowing system certificates, but I may be doing something else wrong.

**Comment by commonsguy** [comment]
I have never used Charles or played much with user certificates. Is there a particular reason it has to be a user certificate, instead of just treating it as a debug-only self-signed certificate?

**Comment by jgilfelt** [comment]
No, not really. It's just that it is uniquely generated for each machine installation of Charles, so doesn't warrant being checked into source control in a team development environment. I'll probably just do as you suggest, then add the file resource path to `.gitignore` so that each developer can override it with their own.

**Comment by commonsguy** [comment]
OK, I implemented what I described in [my first comment](https://github.com/commonsguy/cwac-netsecurity/issues/5#issuecomment-265160401). This will roll out in the next minor release.

**Comment by commonsguy** [comment]
Released with v0.3.0.
